File: app/src/main/java/com/tifone/tnews/home/HomeListPresenter.java
package com.tifone.tnews.home;

import android.annotation.SuppressLint;
import android.text.TextUtils;
import android.util.Log;

import com.google.gson.Gson;
import com.tifone.tnews.api.IMobileNewsApi;
import com.tifone.tnews.bean.news.MultiNewsArticleBean;
import com.tifone.tnews.bean.news.MultiNewsArticleDataBean;
import com.tifone.tnews.home.IHomeContrast.IHomePresenter;
import com.tifone.tnews.home.IHomeContrast.IHomeView;
import com.tifone.tnews.model.NetworkModel;
import com.tifone.tnews.utils.RetrofitFactory;
import com.tifone.tnews.utils.TimeUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by tongkao.chen on 2018/5/4.
 */

public class HomeListPresenter implements IHomePresenter {
    private IHomeView mView;
    private NetworkModel mModel;
    private Random random;
    private String mCategory;
    private String mTime;
    private List<MultiNewsArticleDataBean> mDataList;
    private Gson mGson;
    private List<String> mHistoryItemsId;


    public static HomeListPresenter newInstance(IHomeView view) {
        return new HomeListPresenter(view);
    }

    private HomeListPresenter(IHomeView view) {
        mView = view;
        //mModel = NetworkModel.getInstance();
        random = new Random();
        mTime = TimeUtils.getCurrentTimeStamp();
        mDataList = new ArrayList<>();
        mHistoryItemsId = new ArrayList<>();
        mGson = new Gson();
    }

    @SuppressLint("CheckResult")
    @Override
    public void loadData(String... category) {

        if (null == mCategory) {
            mCategory = category[0];
        }
        if (mView.isLoadLatest()) {
            mTime = TimeUtils.getCurrentTimeStamp();
        } else {
            mTime = String.valueOf(mDataList.get(mDataList.size() - 1).getBehot_time());
        }
        getRandom()
                .subscribeOn(Schedulers.io())
                .switchMap(new Function<MultiNewsArticleBean, Observable<MultiNewsArticleDataBean>>() {

                    @Override
                    public Observable<MultiNewsArticleDataBean> apply(MultiNewsArticleBean multiNewsArticleBean) throws Exception {
                        // 得到新闻的data数据，其包含有新闻的标题、关键字分类信息，同时包含新闻详情的入口
                        List<MultiNewsArticleDataBean> newsDataList = new ArrayList<>();
                        List<MultiNewsArticleBean.DataBean> originalDataBeans = multiNewsArticleBean.getData();
                        for (MultiNewsArticleBean.DataBean dataBean : originalDataBeans) {
                            // 解析Data的json内容
                            MultiNewsArticleDataBean bean = mGson.fromJson(dataBean.getContent(), MultiNewsArticleDataBean.class);
                            if (bean == null || TextUtils.isEmpty(bean.getTitle())) {
                                continue;
                            }
                            newsDataList.add(bean);
                        }
                        return Observable.fromIterable(newsDataList);
                    }
                })
                .filter(new Predicate<MultiNewsArticleDataBean>() {
                    @Override
                    public boolean test(MultiNewsArticleDataBean bean) throws Exception {
                        // 移除已有的item
                        StringBuilder key = new StringBuilder("");
                        key.append(bean.getGroup_id()).
                                append(bean.getItem_id()).
                                append(bean.getItem_version());
                        if (mHistoryItemsId.contains(key.toString())){
                            return false;
                        }
                        mHistoryItemsId.add(key.toString());
                        return true;
                    }
                })
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<MultiNewsArticleDataBean>>() {
                    @Override
                    public void accept(List<MultiNewsArticleDataBean> multiNewsArticleDataBeans) throws Exception {
                        doSetAdapter(multiNewsArticleDataBeans);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        Log.e("tifone", "Get data error: ", throwable);
                    }
                });
    }

    @Override
    public void doSetAdapter(List<MultiNewsArticleDataBean> list) {
        // 数据list已变更，更新ui
        mDataList.addAll(list);
        if (mView.isLoadLatest()) {
            //removeRepeatDataAndAdd(list);
            // 根据时间排序
            Comparator<MultiNewsArticleDataBean> comparator = new Comparator<MultiNewsArticleDataBean>() {
                @Override
                public int compare(MultiNewsArticleDataBean bean1, MultiNewsArticleDataBean bean2) {
                    return bean2.getBehot_time() - bean1.getBehot_time();
                }
            };
            Collections.sort(mDataList, comparator);
        }
        mView.onSetAdapter(mDataList);
    }

    private void removeRepeatDataAndAdd(List<MultiNewsArticleDataBean> list) {

        Set<MultiNewsArticleDataBean> set = new HashSet<>();
        set.addAll(mDataList);
        set.addAll(list);
        List<MultiNewsArticleDataBean> newList = new ArrayList<>();
        newList.addAll(set);
        mDataList.clear();
        mDataList = newList;
    }

    private Observable<MultiNewsArticleBean> getRandom() {
        int i = random.nextInt(10);
        if (i % 2 == 0) {
            return RetrofitFactory.getRetrofit().create(IMobileNewsApi.class).getNewsArticle(mCategory, mTime);
        }
        return RetrofitFactory.getRetrofit().create(IMobileNewsApi.class).getNewsArticle2(mCategory, mTime);
    }
}

File: app/src/main/kotlin/com.tifone.tnews/home/HomeListPresenter.kt
package com.tifone.tnews.home

import android.text.TextUtils
import android.util.Log
import com.google.gson.Gson
import com.tifone.tnews.api.IMobileNewsApi
import com.tifone.tnews.bean.news.MultiNewsArticleBean
import com.tifone.tnews.bean.news.MultiNewsArticleDataBean
import com.tifone.tnews.model.NetworkModel
import com.tifone.tnews.utils.RetrofitFactory
import com.tifone.tnews.utils.TimeUtils
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.lang.Comparable
import java.lang.StringBuilder
import java.util.*

class HomeListPresenter(view: IHomeContrast.IHomeView) : IHomeContrast.IHomePresenter {
    private var mView: IHomeContrast.IHomeView = view
    private var random: Random = Random()
    private var mCategory: String? = null
    private var mTime: String = TimeUtils.getCurrentTimeStamp()
    private var mDataList: MutableList<MultiNewsArticleDataBean> = mutableListOf()
    private var mGson: Gson = Gson()
    private var mHistoryItemsId: MutableList<String> = mutableListOf()

    companion object STATIC {
        fun newInstance(view: IHomeContrast.IHomeView): HomeListPresenter {
            return HomeListPresenter(view)
        }
    }

    override fun loadData(vararg category: String) {
        if (mCategory == null) {
            mCategory = category[0]
        }
        mTime = if (mView.isLoadLatest()) {
            TimeUtils.getCurrentTimeStamp()
        } else {
            mDataList[mDataList.lastIndex].behot_time.toString()
        }
        getRandom().subscribeOn(Schedulers.io())
                .switchMap {
                    val dataList: MutableList<MultiNewsArticleBean.DataBean> = it.data
                    val newDataList: MutableList<MultiNewsArticleDataBean> = mutableListOf()
                    for (dataBean in dataList) {
                        val bean: MultiNewsArticleDataBean = mGson.fromJson(
                                dataBean.content, MultiNewsArticleDataBean::class.java)
                        if (TextUtils.isEmpty(bean.title)) {
                            continue
                        }
                        newDataList.add(bean)
                    }
                    // 返回
                    Observable.fromIterable(newDataList)
                }
                .filter {
                    val key = StringBuilder("")
                    key.append(it.group_id)
                            .append(it.item_id)
                            .append(it.item_version)
                    // 是否已经加载过，返回值
                    key.toString() !in mHistoryItemsId
                }
                .toList()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    // consumer
                    doSetAdapter(it)
                }, {
                    // throwable
                    Log.e("tifone", "Get data error: ", it)
                })
    }

    override fun doSetAdapter(list: MutableList<MultiNewsArticleDataBean>) {
        mDataList.addAll(list)
        if (mView.isLoadLatest()) {
            val comparator: Comparator<MultiNewsArticleDataBean> = Comparator {
                bean1, bean2 ->
                // 对比规则
                bean2.behot_time - bean1.behot_time
            }
            Collections.sort(mDataList, comparator)
        }
        mView.onSetAdapter(mDataList)
    }

    private fun getRandom(): Observable<MultiNewsArticleBean> {
        return if (random.nextInt(10) % 2 == 0) {
            RetrofitFactory.getRetrofit().create(IMobileNewsApi::class.java).getNewsArticle(mCategory, mTime)
        } else {
            RetrofitFactory.getRetrofit().create(IMobileNewsApi::class.java).getNewsArticle2(mCategory, mTime)
        }
    }
}